src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+        try f64_hash_map.put(rand_number, rand_number);
+        try f64_hash_map.put(rand_number - deviation_margin, rand_number - deviation_margin);
+        try f64_hash_map.put(rand_number + deviation_margin, rand_number + deviation_margin);
+    }
+
+    // All elements are expected to be hashed independently, without collisions.
+    try testing.expectEqual(300, f64_hash_map.count());
+
+    var iterator = f64_hash_map.iterator();
+    while (iterator.next()) |entry| {
+        const key = entry.key_ptr.*;
+        const value = entry.value_ptr.*;
+        try testing.expectEqual(key, value);
+    }
+}
+
+test "hashmap can map f64 to segment metadata array list" {
+    const allocator = testing.allocator;
+    var f64_metadata_hash_map = HashMap(
+        f64,
+        ArrayList(SegmentMetadata),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ).init(allocator);
+    defer {
+        var iterator = f64_metadata_hash_map.iterator();
+        while (iterator.next()) |entry| {
+            entry.value_ptr.*.deinit();
+        }
+        f64_metadata_hash_map.deinit();
+    }
+
+    var f64_usize_hash_map = HashMap(
+        f64,
+        usize,
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ).init(allocator);
+    defer f64_usize_hash_map.deinit();
+
+    var rnd = std.rand.DefaultPrng.init(@as(u64, @bitCast(std.time.milliTimestamp())));
+
+    for (0..200) |_| {
+        const rand_number = @floor((rnd.random().float(f64) - 0.5) * 100) / 10;
+
+        const count_map_result = try f64_usize_hash_map.getOrPut(rand_number);
+        if (!count_map_result.found_existing) {
+            count_map_result.value_ptr.* = 0;
+        }
+        count_map_result.value_ptr.* += 1;
+
+        const metadata_map_result = try f64_metadata_hash_map.getOrPut(rand_number);
+        if (!metadata_map_result.found_existing) {
+            metadata_map_result.value_ptr.* = ArrayList(SegmentMetadata).init(allocator);
+        }
+        try metadata_map_result.value_ptr.*.append(SegmentMetadata{
+            .start_time = count_map_result.value_ptr.*,
+            .interception = rand_number,
+            .lower_bound_slope = rand_number,
+            .upper_bound_slope = rand_number,
+        });
+    }
+
+    var iterator_map = f64_metadata_hash_map.iterator();
+    while (iterator_map.next()) |entry| {
+        const expected_array_size: usize = f64_usize_hash_map.get(entry.key_ptr.*).?;
+        try testing.expectEqual(expected_array_size, entry.value_ptr.*.items.len);
+
+        for (entry.value_ptr.*.items, 1..) |item, i| {
+            try testing.expectEqual(i, item.start_time);
+            try testing.expectEqual(entry.key_ptr.*, item.lower_bound_slope);
+        }
+    }
+}
+
+test "sim-piece can compress, decompress and merge many segments with non-zero error bound" {
+    const allocator = testing.allocator;
+
+    var list_values = ArrayList(f64).init(allocator);
```

Comment:
The purpose of `list_values` is unclear for most of the test. Maybe rename to `uncompressed_values` and convert to slice when calling `compress()` so `list_values` can be removed.

Reply:
Fixed.

Reply:
Is `[0..]` not only needed if a sub-slice is needed? Does the simpler `uncompressed_values.items` or `&uncompressed_values.items` not work?

Reply:
Fixed.